File: internal/compute/repo.go
package compute

import (
	"context"
	"encoding/json"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"

	"pii/internal/models"
)

const repoCacheTTL = 2 * time.Second

type repoCacheEntry struct {
	config models.SystemConfig
	exp    time.Time
}

// Repo manages models.SystemConfig in Redis with a short TTL cache.
type Repo struct {
	rdb   *redis.Client
	ns    string
	mu    sync.Mutex
	cache map[string]repoCacheEntry
}

// NewRepo creates a system config repo.
func NewRepo(rdb *redis.Client, ns string) *Repo {
	return &Repo{
		rdb:   rdb,
		ns:    ns,
		cache: map[string]repoCacheEntry{},
	}
}

// SaveSystem stores a system config and bumps the config epoch.
func (r *Repo) SaveSystem(ctx context.Context, config models.SystemConfig) error {
	key := systemKey(r.ns, config.SystemID)
	raw, err := json.Marshal(config)
	if err != nil {
		return err
	}
	pipe := r.rdb.TxPipeline()
	pipe.Set(ctx, key, raw, 0)
	pipe.SAdd(ctx, systemsSet(r.ns), config.SystemID)
	pipe.Incr(ctx, configEpochKey(r.ns))
	if _, err := pipe.Exec(ctx); err != nil {
		return err
	}
	r.mu.Lock()
	r.cache[config.SystemID] = repoCacheEntry{config: config, exp: time.Now().Add(repoCacheTTL)}
	r.mu.Unlock()
	return nil
}

// GetSystem returns a system config using the cache.
func (r *Repo) GetSystem(ctx context.Context, systemID string) (*models.SystemConfig, error) {
	r.mu.Lock()
	if e, ok := r.cache[systemID]; ok && time.Now().Before(e.exp) {
		cfg := e.config
		r.mu.Unlock()
		return &cfg, nil
	}
	r.mu.Unlock()

	key := systemKey(r.ns, systemID)
	raw, err := r.rdb.Get(ctx, key).Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var cfg models.SystemConfig
	if err := json.Unmarshal([]byte(raw), &cfg); err != nil {
		return nil, err
	}
	r.mu.Lock()
	r.cache[systemID] = repoCacheEntry{config: cfg, exp: time.Now().Add(repoCacheTTL)}
	r.mu.Unlock()
	return &cfg, nil
}

// ListSystems returns all registered system configs.
func (r *Repo) ListSystems(ctx context.Context) ([]models.SystemConfig, error) {
	ids, err := r.rdb.SMembers(ctx, systemsSet(r.ns)).Result()
	if err != nil {
		return nil, err
	}
	var out []models.SystemConfig
	for _, id := range ids {
		cfg, err := r.GetSystem(ctx, id)
		if err != nil {
			return nil, err
		}
		if cfg != nil {
			out = append(out, *cfg)
		}
	}
	return out, nil
}

// BumpConfigEpoch increments the config epoch.
func (r *Repo) BumpConfigEpoch(ctx context.Context) error {
	return r.rdb.Incr(ctx, configEpochKey(r.ns)).Err()
}

// GetControlEpoch returns the current config epoch.
func (r *Repo) GetControlEpoch(ctx context.Context) (int64, error) {
	return r.rdb.Get(ctx, configEpochKey(r.ns)).Int64()
}

File: compute/src/repo.rs
//! Репозиторий конфигураций систем-потребителей.
//!
//! Хранит [`SystemConfig`] в Redis и кэширует их in-memory на короткое
//! время (2 секунды), чтобы снизить нагрузку на Redis при частых чтениях.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use redis::aio::ConnectionManager;
use redis::AsyncCommands;

use crate::keys::{config_epoch_key, system_key, systems_set};
use crate::models::SystemConfig;

/// Время жизни in-memory кэша конфигов.
const CACHE_TTL: Duration = Duration::from_secs(2);

/// Внутреннее состояние кэша конфигов.
struct SystemCache {
    /// Карта «system_id → (конфиг, время записи)».
    entries: HashMap<String, (SystemConfig, Instant)>,
}

impl SystemCache {
    /// Создаёт пустой кэш.
    fn new() -> Self {
        SystemCache {
            entries: HashMap::new(),
        }
    }

    /// Возвращает конфиг, если он есть в кэше и ещё не истёк.
    fn get(&self, system_id: &str) -> Option<SystemConfig> {
        let (config, written_at) = self.entries.get(system_id)?;
        if written_at.elapsed() > CACHE_TTL {
            return None;
        }
        Some(config.clone())
    }

    /// Кладёт конфиг в кэш с текущей временной меткой.
    fn put(&mut self, system_id: &str, config: SystemConfig) {
        self.entries
            .insert(system_id.to_string(), (config, Instant::now()));
    }

    /// Очищает кэш (вызывается при сохранении конфига).
    fn invalidate(&mut self) {
        self.entries.clear();
    }
}

/// Репозиторий конфигураций систем.
#[derive(Clone)]
pub struct Repo {
    /// Менеджер подключения к Redis.
    conn: ConnectionManager,
    /// Пространство имён ключей.
    ns: String,
    /// In-memory кэш конфигов.
    cache: Arc<Mutex<SystemCache>>,
}

impl Repo {
    /// Создаёт репозиторий поверх подключения к Redis.
    pub fn new(conn: ConnectionManager, ns: String) -> Self {
        Repo {
            conn,
            ns,
            cache: Arc::new(Mutex::new(SystemCache::new())),
        }
    }

    /// Сохраняет конфиг системы в Redis и инвалидирует кэш.
    ///
    /// Запись выполняется транзакционно: конфиг кладётся по ключу системы,
    /// а идентификатор добавляется во множество всех систем.
    pub async fn save_system(&self, config: &SystemConfig) -> redis::RedisResult<()> {
        let mut conn = self.conn.clone();
        let key = system_key(&self.ns, &config.system_id);
        let set = systems_set(&self.ns);
        let json = serde_json::to_string(config).map_err(serde_err)?;
        redis::pipe()
            .set(&key, json)
            .sadd(&set, &config.system_id)
            .query_async::<_, ()>(&mut conn)
            .await?;
        self.cache.lock().unwrap().invalidate();
        Ok(())
    }

    /// Возвращает конфиг системы, используя in-memory кэш.
    ///
    /// Сначала проверяется кэш; при промахе конфиг читается из Redis и
    /// кладётся в кэш.
    pub async fn get_system(&self, system_id: &str) -> redis::RedisResult<Option<SystemConfig>> {
        if let Some(config) = self.cache.lock().unwrap().get(system_id) {
            return Ok(Some(config));
        }
        let mut conn = self.conn.clone();
        let key = system_key(&self.ns, system_id);
        let raw: Option<String> = conn.get(&key).await?;
        let config = match raw {
            Some(json) => {
                let config: SystemConfig = serde_json::from_str(&json).map_err(serde_err)?;
                self.cache.lock().unwrap().put(system_id, config.clone());
                Some(config)
            }
            None => None,
        };
        Ok(config)
    }

    /// Возвращает список всех зарегистрированных систем.
    pub async fn list_systems(&self) -> redis::RedisResult<Vec<SystemConfig>> {
        let mut conn = self.conn.clone();
        let set = systems_set(&self.ns);
        let ids: Vec<String> = conn.smembers(&set).await?;
        let mut systems = Vec::with_capacity(ids.len());
        for id in ids {
            if let Some(config) = self.get_system(&id).await? {
                systems.push(config);
            }
        }
        Ok(systems)
    }

    /// Увеличивает эпоху конфигурации (сигнал о смене конфигов).
    pub async fn bump_config_epoch(&self) -> redis::RedisResult<()> {
        let mut conn = self.conn.clone();
        let key = config_epoch_key(&self.ns);
        conn.incr::<_, _, ()>(&key, 1).await?;
        Ok(())
    }

    /// Возвращает текущее значение эпохи конфигурации.
    pub async fn get_control_epoch(&self) -> redis::RedisResult<i64> {
        let mut conn = self.conn.clone();
        let key = config_epoch_key(&self.ns);
        let value: i64 = conn.get(&key).await.unwrap_or(0);
        Ok(value)
    }
}

/// Преобразует ошибку сериализации в ошибку Redis.
fn serde_err(e: serde_json::Error) -> redis::RedisError {
    redis::RedisError::from((redis::ErrorKind::ResponseError, "serde", e.to_string()))
}
